connman/src/storage.c

```diff
@@ -794,11 +796,58 @@ static void keyfile_cleanup(void)
 	keyfile_hash = NULL;
 }
 
+static int validate_path(const char *pathname)
+{
+	char *expanded_path;
+	int err = 0;
+
+	/* realpath() expands all symbolic links and references. */
+	expanded_path = realpath(pathname, NULL);
+	if (!expanded_path) {
+		switch (errno) {
+		case EACCES:
+		case EINVAL: /* pathname was NULL */
+		case EIO:
+		case ELOOP:
+		case ENAMETOOLONG:
+		case ENOMEM:
+		case ENOTDIR:
+			DBG("realpath() error %d/%s", errno, strerror(errno));
+			return -errno;
+		/* When saving the file does not exist */
+		case ENOENT:
+			return 0;
+		}
+
+		return -errno;
+	}
+
+	/*
+	 * If the paths differ the path most likely contains symlink or is not
+	 * an absolute path.
+	 */
+	if (g_strcmp0(pathname, expanded_path))
+		err = -ENOTSUP;
+
+	g_free(expanded_path);
+
+	return err;
+
+}
+
 static GKeyFile *storage_load(const char *pathname)
 {
 	struct keyfile_record *record = NULL;
 	GError *error = NULL;
 	GKeyFile *keyfile = NULL;
+	int err;
+
+	err = validate_path(pathname);
```

Comment:
Does it make sense to do this before the cache check?

Reply:
Maybe not now that you brought that up. Late Friday changes are at first like this. It is indeed faster and one allocation less in cases where the service-files etc. are not user specific as this issue exist only when multiuser support is in use.